### PR TITLE
Provide a results service

### PR DESCRIPTION
**Pull request by steelbrain**
Fixes AtomLinter/linter-plus#4
/cc @abe33 @hd-deman @iam4x

### CODE REVIEW

File: lib/main.coffee [hunk 1]
```diff
@@ -15,5 +15,8 @@ module.exports =
   consumeLinter: (Linter) ->
     @Instance.Linters.push Linter
 
+  provideResults: ->
+    @Instance
+
```
Comment (iam4x) [inline lib/main.coffee]: Shouldn't we give only the `@Instance.Messages` as return?

File: lib/linter.coffee [hunk 2]
```diff
@@ -103,6 +108,9 @@ class Linter
       @View.update()
       @ViewPanel.show() unless @ViewPanel.isVisible()
 
+  onDidUpdate:(Callback)->
```
Comment (iam4x) [inline lib/linter.coffee]: `onDidUpdate: (Callback) ->`

(respect coffeelint.json code style)

File: lib/main.coffee (hunk 1 above)
Comment (steelbrain) [inline lib/main.coffee]: Nope, 'cause @Messages is an array not an object, Consumers can then do
```js
(Instance)->
  Instance.onDidUpdate(Messages)->
```

File: lib/linter.coffee (hunk 2 above)
Comment (steelbrain) [inline lib/linter.coffee]: I don't really use Atom for writing Atom Packages :P I use phpStorm and it looked good to it. Gonna fix now.

File: lib/main.coffee (hunk 1 above)
Comment (iam4x) [inline lib/main.coffee]: So it's more `provideLinterPlus`, and then you can use every functions on the linter?

Comment (steelbrain) [inline lib/main.coffee]: Well, Yeah.
It's like that so when we add new `onDid` things in the future, we won't have to anything.

Comment (iam4x) [inline lib/main.coffee]: Cool, just change names so :)

**Comment by iam4x** [comment]
Just add something on the `README.md` for documentation and we're good.

**Comment by steelbrain** [comment]
I am not updating the `README.md` on purpose, because almost everything on it is out of date, the classes have been renamed, the installation method doesn't work and stuff.
so I say, we add this documentation when we fix the other stuff. What is your opinion?

**Comment by iam4x** [comment]
We can start now, just add a little note (at least to remember for after).

Every PR should now modify `README.md` with updated doc :+1:

**Comment by steelbrain** [comment]
I just found a bug in it, atom services don't treat dots like dashes, returning something from `linter.plus.self` would be given to our own `linter.plus` in an object like `{self: Linter}`

**Comment by iam4x** [comment]
Good catch, I don't really know the `consumedServices` / `providedServices` API

**Comment by steelbrain** [comment]
Please note that you will have to update the `plus` linters to match this new name.

**Comment by iam4x** [comment]
Alright, do you have anything to add?

**Comment by steelbrain** [comment]
The documentation of the new API, but I am busy with another project right now so will do it after like 10 hours or sooner maybe.

**Comment by abe33** [comment]
In an ideal API, I would expect the linter package to provides the following features: 

- A wrapper model (say `EditorLinter`) for each editor that has a linter enabled, the model is mostly the binding between the underlying linter and the editor. They are created in a `observeTextEditors` callback only if there's a linter available for the editor's file and should be stored by the package for later use.
- This model is the one onto which register to be notified of update for the linter results of the given editor (using a `onDidUpdateResults` method).
- These models should be available through various methods (`getEditorLinters`, `getActiveEditorLinter`, `observeEditorLinters`, `onDidAddEditorLinter`, `onDidRemoveEditorLinter`, etc.) exposed on the provided service object.

The purpose of the binding model is to have a clean interface to be notified of one editor results. It also allow to keep a track of the previous results and optimize updates (both internally in the linter or externally in a package that consumes the linter API).

What do you think of this?

**Comment by steelbrain** [comment]
I personally don't think that we'll need that complexity when all we're trying to do is another package the status of errors in the code.
We will possibly have lots of linters for one single editor, and sometimes no linter all for an editor, I also fail to understand how `observeEditorLinters` and `onDidAddEditerLinter` is gonna benit the api consumer.
Here's how a package like minimap can use linter's result in the API provided.
```js
module.exports = 
  linterDecorations:[]
  consumeLinter:(Linter)->
    Linter.onDidUpdate (Messages)->
      @removeLinterDecorations()
      ActivePath = atom.workspace.getActiveEditor().getPath()
      Messages = Messages.filter (Entry) -> Entry.File === ActivePath
      @decorateLinter(Messages)
```
that Messages array here contains Objects that have `Message:String`, `File:String`, `Position:Range`  and sometimes `Trace:array<LinterTrace>` attributes.

**Comment by abe33** [comment]
If I take your example, the only thing I receive are the messages, so I'll have to guess the correct editor from them, if there's any. 
Now let's introduce come concurency, what if I change the active editor while the event is being dispatched with no messages? I no longer can attach the errors to the right editor. 

Also, following your example, whatever number of panes I have open there's only one minimap that'll display linter errors, and when new results are dispatched the previous editor's minimap will lose its own errors (always due to relying on the active pane). I would expect a more coherent approach.

> I personally don't think that we'll need that complexity 

Basically you're just pushing the complexity on the consumers. 
For a basic use case like *I want to display the errors on a per-editor basis*, I'll have to activate every editors, trigger an update to have the linter executed, retrieve the results, and binding them to a path (which is not possible when there's no results). A consumer shouldn't have to care about the logic of how linters works, it should only be concerned about the results: which editor, which results.

> We will possibly have lots of linters for one single editor, and sometimes no linter all for an editor

No problem here, the binding model can have a 1-n cardinality with linters. And if there's no linter then there's no binding model instance to create.

> I also fail to understand how observeEditorLinters and onDidAddEditerLinter is gonna benit the api consumer.

A consumer will be able to register to every existing and future models, in a way that is coherent with the rest of Atom's API. For instance, if I were to implement a minimap plugin, I'll register a callback to `observeEditorLinters` and then create a binding model between the editor's minimap and the EditorLinter model so that ever time new results are dispatched I can update the concerned minimap without having to care about the others. And I don't have to care to whether the editor have linters or not, if I get a model then there's linters. I don't have to care about the end of life of these objects, when the editor is destroy so is the minimap, and I can just listen to that.

I think the purpose of an API should be to enable a third party to implement features that were not originally anticipated. And a clean domain model is essential to do that.

**Comment by steelbrain** [comment]
Thank you for proving me wrong, You are right about the confusion of editors and linters. I'll update the API to reflect the changes that you suggested.

**Comment by steelbrain** [comment]
I am just gonna close this PR and open up a new one.

**Pull request closed without merging**